src/assign.c

```diff
@@ -152,13 +152,24 @@ static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
   R_len_t i,l;
   int protecti=0;
   SEXP newdt = PROTECT(allocVector(VECSXP, n)); protecti++;   // to do, use growVector here?
-  //copyMostAttrib(dt, newdt);   // including class
-  DUPLICATE_ATTRIB(newdt, dt);
+  SET_ATTRIB(newdt, shallow_duplicate(ATTRIB(dt)));
+  SET_OBJECT(newdt, OBJECT(dt));
+  //SHALLOW_DUPLICATE_ATTRIB(newdt, dt);  // SHALLOW_DUPLICATE_ATTRIB would be a bit neater but is only available from R 3.3.0
+  
   // TO DO: keepattr() would be faster, but can't because shallow isn't merely a shallow copy. It
   //        also increases truelength. Perhaps make that distinction, then, and split out, but marked
   //        so that the next change knows to duplicate.
-  //        Does copyMostAttrib duplicate each attrib or does it point? It seems to point, hence DUPLICATE_ATTRIB
-  //        for now otherwise example(merge.data.table) fails (since attr(d4,"sorted") gets written by setnames).
+  //        keepattr() also merely points to the entire attrbutes list and thus doesn't allow replacing
+  //        some of its elements.
+  
+  // We copy all attributes that refer to column names so that calling setnames on either
+  // the original or the shallow copy doesn't break anything.
+  SEXP index = PROTECT(getAttrib(dt, sym_index)); protecti++;
+  setAttrib(newdt, sym_index, shallow_duplicate(index));
+  
+  SEXP sorted = PROTECT(getAttrib(dt, sym_sorted)); protecti++;
+  setAttrib(newdt, sym_sorted, duplicate(sorted));
```

Comment:
any reason to use deep copy here if we're doing shallow copy for the other attributes?

Reply:
This is just character vector `c("key_col1","key_col2")`. I don't think it cost anything

Reply:
They are exactly the same in this situation. Both make a copy of the atomic vector, including a string vector and only differ in how they treat the attributes. This seemed to me to be the more defensive approach.

Reply:
works for me! thanks

Reply:
This absolutely needs to be a deep copy, iiuc. That's what a comment in the code was trying to convey. Consider when a user uses `setnames()` to change a column's name, and that column is involved in a key or index. `setnames` looks to see if that column is involved in the key or any indexes, and if so updates the corresponding name in the key or index name. If it wasn't a deep copy of names, that change-of-column-name in the copy would break the original table.  Better is to always shallow copy and put a flag on it so that the next update knows to deep copy, however we didn't get to that yet.  Plus full reference counting in R came along which we may be able to tie in with rather than do our own thing.